sproxy.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sproxy
+
+import (
+	"errors"
+	"sync"
+	"time"
+)
+
+type Event struct {
+	StatusCode   int       `json:"code,omitempty"`
+	Request      bool      `json:"is_req,omitempty"`
+	ResponseSize int64     `json:"res_size,omitempty"`
+	StartTime    time.Time `json:"start,omitempty"`
+	EndTime      time.Time `json:"end,omitempty"`
+	ID           string    `json:"id,omitempty"`
+	Err          error     `json:"err,omitempty"`
+}
+
+type MetricsReporter struct {
+	currentEvents []*Event
+	period        time.Duration
+
+	cancelChan <-chan bool
+	cancelFn   func() error
+
+	// mu protects statsRecvFn
+	mu          sync.RWMutex
+	statsRecvFn func(*InflightStats) error
+}
+
+const defaultPeriod = 1 * time.Minute
+
+func NewMetricsReporter(period time.Duration) (*MetricsReporter, error) {
+	if period <= 0 {
+		period = defaultPeriod
+	}
+	cancelChan, cancelFn := makeCanceler()
+	mr := &MetricsReporter{
+		period:        period,
+		currentEvents: make([]*Event, 0, 512), // Arbitrary value
+		cancelChan:    cancelChan,
+		cancelFn:      cancelFn,
+	}
+	return mr, nil
+}
+
+func (mr *MetricsReporter) Close() error {
+	if mr.cancelFn == nil {
+		return nil
+	}
+	return mr.cancelFn()
+}
+
+func (mr *MetricsReporter) AddEvent(evt *Event) {
+	mr.mu.Lock()
+	mr.currentEvents = append(mr.currentEvents, evt)
+	mr.mu.Unlock()
+}
+
+var errAlreadyClosed = errors.New("already closed")
+
+func makeCanceler() (<-chan bool, func() error) {
+	cancelChan := make(chan bool, 1)
+	var cancelOnce sync.Once
+	cancelFn := func() error {
+		var err error = errAlreadyClosed
+		cancelOnce.Do(func() {
+			close(cancelChan)
+			err = nil
+		})
+		return err
+	}
+
+	return cancelChan, cancelFn
+}
+
+// Do is a synchronous method that reports events
+// at the end of every specified sampling period
+// sending them to the StackDriver Monitoring API.
+func (mr *MetricsReporter) Do() {
+	startTime := time.Now()
+	i := uint64(0)
+	for {
+		select {
+		case <-mr.cancelChan:
+			return
+		case <-time.After(mr.period):
+			// Otherwise the duration has lapsed
+		}
+
+		i += 1
+		drainedEvents := mr.drainEvents()
+
+		go mr.reportEvents(drainedEvents, startTime, time.Now(), i)
```

Comment:
shouldn't startTime be in the for loop? Isn't it the start time of each period?

Reply:
It was meant to be the global startTime i.e to the time since we started tracking but we aren't necessarily using that right now, so I'll scrap it and in the future if we need it we can use it.